File: .claude/hooks/block-git-add.py
"""PreToolUse hook: block broad git add commands (git add . / git add -A)."""
import json
import sys

def main():
    input_data = json.loads(sys.stdin.read())
    tool_input = input_data.get("tool_input", {})
    command = tool_input.get("command", "")

    blocked_patterns = ["git add .", "git add -A", "git add --all"]
    for pattern in blocked_patterns:
        if pattern in command:
            print(json.dumps({
                "decision": "block",
                "reason": f"Blocked broad staging command '{pattern}'. Stage specific files instead."
            }))
            return

    print(json.dumps({"decision": "approve"}))

if __name__ == "__main__":
    main()

File: .claude/hooks/block-git-add.js
#!/usr/bin/env node
/** PreToolUse hook: block broad git add commands (git add . / git add -A). */

let input = '';
process.stdin.on('data', chunk => { input += chunk; });
process.stdin.on('end', () => {
  try {
    const data = JSON.parse(input);
    const command = (data.tool_input || {}).command || '';
    const blocked = ['git add .', 'git add -A', 'git add --all'];
    for (const pattern of blocked) {
      if (command.includes(pattern)) {
        console.log(JSON.stringify({
          decision: 'block',
          reason: `Blocked broad staging command '${pattern}'. Stage specific files instead.`
        }));
        process.exit(0);
      }
    }
    console.log(JSON.stringify({ decision: 'approve' }));
  } catch {
    console.log(JSON.stringify({ decision: 'approve' }));
  }
});
